services/skus/playstore.go

```diff
@@ -31,3 +44,297 @@ func (x *playStoreSubPurchase) isPending() bool {
 
 	return state == pending || state == pendingDef
 }
+
+type gpsValidatorConfig struct {
+	aud      string
+	iss      string
+	svcAcct  string
+	disabled bool
+}
+
+type gpsTokenValidator interface {
+	Validate(ctx context.Context, token, aud string) (*idtoken.Payload, error)
+}
+
+type gpsNtfAuthenticator struct {
+	cfg   gpsValidatorConfig
+	valid gpsTokenValidator
+}
+
+func newGPSNtfAuthenticator(cfg gpsValidatorConfig, valid gpsTokenValidator) *gpsNtfAuthenticator {
+	result := &gpsNtfAuthenticator{
+		cfg:   cfg,
+		valid: valid,
+	}
+
+	return result
+}
+
+func (x *gpsNtfAuthenticator) authenticate(ctx context.Context, hdr string) error {
+	if x.cfg.disabled {
+		return errGPSDisabled
+	}
+
+	if hdr == "" {
+		return errGPSAuthHeaderEmpty
+	}
+
+	token := strings.Split(hdr, " ")
+	if len(token) != 2 {
+		return errGPSAuthHeaderFmt
+	}
+
+	p, err := x.valid.Validate(ctx, token[1], x.cfg.aud)
+	if err != nil {
+		return fmt.Errorf("invalid authentication token: %w", err)
+	}
+
+	if p.Issuer == "" || p.Issuer != x.cfg.iss {
+		return errGPSInvalidIssuer
+	}
+
+	if p.Claims["email"] != x.cfg.svcAcct {
+		return errGPSInvalidEmail
+	}
+
+	if p.Claims["email_verified"] != true {
+		return errGPSEmailNotVerified
+	}
+
+	return nil
+}
+
+// playStoreDevNotification represents a notification from Play Store.
+//
+// For details, see https://developer.android.com/google/play/billing/rtdn-reference.
+type playStoreDevNotification struct {
+	PackageName       string                      `json:"packageName"`
+	EventTimeMilli    json.Number                 `json:"eventTimeMillis"`
+	SubscriptionNtf   *playStoreSubscriptionNtf   `json:"subscriptionNotification"`
+	VoidedPurchaseNtf *playStoreVoidedPurchaseNtf `json:"voidedPurchaseNotification"`
+
+	// Only presense of these matters. The content is ignored.
+	OneTimeProductNtf *struct{} `json:"oneTimeProductNotification"`
```

Comment:
why does the presence matter? unmarshalling shouldn't fail if these were omitted completely

Reply:
Look at the code in the methods on this type – based on the presence of these fields the type of the notification is is determined.

If these fields were omitted entirely, it would not be possible to determine some types of notifications – specifically the ones which would be omitted. It would not be possible to determine if it were a test or one time product notification.

By making the fields `*struct{}` the goal is accomplished – the type of notification is determined, but the content is discarded.